File: othersAlgorithm/link_list.py
from random import randint


class Node(object):
    def __init__(self, value):
        self.value = value
        self.next = None

    def __str__(self):
        return f'<Node> -> {self.value}'

    __repr__ = __str__


class LinkList(object):
    def __init__(self, nums: list):
        self.head = None
        opt = input("是否逆序构造(y/n)")
        if opt == 'y':
            for item in nums:
                node = Node(item)
                node.next = self.head
                self.head = node
        else:
            p = self.head
            for item in nums:
                node = Node(item)
                if p is None:
                    p = node
                    self.head = node
                else:
                    node.next = p.next
                    p.next = node

    def reverse(self):
        """
        反转链表
        """
        print('Reverse')
        p, q = self.head, self.head.next
        p.next = None
        while q is not None:
            tmp = q.next
            q.next = p
            p = q
            q = tmp
        self.head = p

    def sort(self):
        print('Sort')
        def _sort(head: Node, end=None):
            if head ==end or head.next == end:
                return
            aim = head
            p, q = head, head.next
            while q != end:
                if q.value < aim.value:
                    p = p.next
                    p.value, q.value = q.value, p.value
                q = q.next
            p.value, head.value = head.value, p.value
            _sort(head, p)
            _sort(p.next, end)
        _sort(self.head, None)

    def display(self):
        p = self.head
        while p is not None:
            print(p, end=' ')
            p = p.next
        print()

    def has_cycle(self) -> tuple:
        """
        判断链表是否有环
        :return:
        """
        p, q = self.head, self.head.next
        while q is not None and q.next is not None and q != p:
            q = q.next.next
            p = p.next
        if q != p:
            return False, "doest not has cycle"
        tmp = self.head
        while tmp != p:
            tmp = tmp.next
            p = p.next
        return tmp, "入口"


if __name__ == '__main__':
    ll = LinkList([randint(1, 30) for i in range(randint(3, 10))])
    ll.display()
    ll.reverse()
    ll.display()
    ll.sort()
    ll.display()
    print(ll.has_cycle())

File: othersAlgorithm/go-examples/list/link_list.go
package list

import "fmt"

type Node struct {
	data int
	Next *Node
}

type LinkList struct {
	head *Node
}

func (ll *LinkList) initList(nums []int) {
	fmt.Print("是否反序构造?(y/n)")
	var opt byte
	fmt.Scanf("%c", &opt)
	if opt == 'Y' || opt == 'y' {
		for _, v := range nums {
			tmp := &Node{v, nil}
			tmp.Next = ll.head
			ll.head = tmp
		}
	} else {
		p := ll.head
		for _, v := range nums {
			tmp := &Node{v, nil}
			if p == nil {
				p = tmp
				ll.head = p
			} else {
				tmp.Next = p.Next
				p.Next = tmp
				p = p.Next
			}
		}
	}
}

func (ll *LinkList) Insert(data int) bool {
	if ll.head == nil {
		ll.head = &Node{data, nil}
		return true
	}
	node := &Node{data, nil}
	p := ll.head
	for p.Next != nil {
		p = p.Next
	}
	node.Next = p.Next
	p.Next = node
	return true
}

func (ll *LinkList) Display() {
	p := ll.head
	for p != nil {
		fmt.Print(p.data, " ")
		p = p.Next
	}
	fmt.Println()
}

func (ll *LinkList) Reverse() {
	p, q := ll.head, ll.head.Next
	p.Next = nil
	for q != nil {
		tmp := q.Next
		q.Next = p
		p = q
		q = tmp
	}
	ll.head = p
}

func (ll *LinkList) Sort() {
	// 链表的排序
	sort(ll.head, nil)
}

func sort(head, end *Node) {
	if head == end || head.Next == end {
		return
	}
	aim := head
	p, q := head, head.Next
	for q != nil {
		if q.data < aim.data {
			p = p.Next
			p.data, q.data = q.data, p.data
		}
		q = q.Next
	}
	aim.data, p.data = p.data, aim.data
	sort(head, p)
	sort(p.Next, end)
}

func (ll *LinkList) HasCycle() bool {
	// 判断是否存在环
	p, q := ll.head, ll.head.Next
	for q != nil && q.Next != nil && p != q {
		p = p.Next
		q = q.Next.Next
	}
	if q != p {
		fmt.Println("不存在环: ")
		return false
	} else {
		fmt.Println("存在环, 寻找环")
		p = ll.head
		for p != q {
			p = p.Next
			q = q.Next
		}
		fmt.Print("入口点: ", p)
		return true
	}
}

func (ll *LinkList) MiddleNode() string {
	//如果是一个节点, 就输出一个, 否则输出两个
	p, q := ll.head, ll.head.Next
	// 一个走一步, 一个走两部
	for q != nil && q.Next != nil {
		p = p.Next
		q = q.Next.Next
	}
	if q == nil {
		return fmt.Sprintf("Only One %d\n", p.data)
	} else {
		return fmt.Sprintf("Double %d-%d\n", p.data, p.Next.data)
	}
}
